nixos/modules/services/cluster/patroni/default.nix

```diff
@@ -0,0 +1,249 @@
+      '';
+    };
+
+    environment = mkOption {
+      type = with types; attrsOf (nullOr (oneOf [ str path package ]));
+      default = { };
+      example = {
+        PATRONI_REPLICATION_PASSWORD = "/secret/file";
+        PATRONI_SUPERUSER_PASSWORD = "/secret/file";
+      };
+      description = "Environment variables made available to Patroni, useful for providing secrets from files.";
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    services.patroni.settings = {
+      scope = cfg.scope;
+      name = cfg.name;
+      namespace = cfg.namespace;
+
+      restapi = {
+        listen = "${cfg.nodeIp}:${toString(cfg.restApiPort)}";
+        connect_address = "${cfg.nodeIp}:${toString(cfg.restApiPort)}";
+      };
+
+      raft = {
+        data_dir = "${cfg.dataDir}/raft";
+        self_addr = "${cfg.nodeIp}:5010";
+        partner_addrs = map (ip: ip + ":5010") cfg.otherNodesIps;
+      };
+
+      postgresql = {
+        listen = "${cfg.nodeIp}:${toString(cfg.postgresqlPort)}";
+        connect_address = "${cfg.nodeIp}:${toString(cfg.postgresqlPort)}";
+        data_dir = cfg.postgresqlDataDir;
+        bin_dir = "${postgresql}/bin";
+        pgpass = "${cfg.dataDir}/pgpass";
+      };
+      watchdog = mkIf cfg.watchdog {
+        mode = "required";
+        device = "/dev/watchdog";
+        safety_margin = 5;
+      };
+    };
+
+    users.users.postgres =
+      {
+        name = "postgres";
+        uid = config.ids.uids.postgres;
+        group = "postgres";
+        description = "PostgreSQL server user";
+        home = "${cfg.postgresqlDataDir}";
+        useDefaultShell = true;
+      };
+
+    users.groups.postgres.gid = config.ids.gids.postgres;
+
+    systemd.services = {
+      patroni = {
+        description = "Runners to orchestrate a high-availability PostgreSQL";
+
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+
+        environment = cfg.environment;
+
+        serviceConfig = mkMerge [{
+          User = "postgres";
+          Group = "postgres";
+          Type = "simple";
+          Restart = "on-failure";
+          TimeoutSec = 30;
+          ExecStart = "${patroni}/bin/patroni ${configFile}";
+          ExecReload = "${pkgs.coreutils}/bin/kill -s HUP $MAINPID";
+          KillMode = "process";
+        }
+          (mkIf (cfg.postgresqlDataDir == "/var/lib/postgresql/${postgresql.psqlSchema}" && cfg.dataDir == "/var/lib/patroni") {
+            StateDirectory = "patroni patroni/raft postgresql postgresql/${postgresql.psqlSchema}";
+            StateDirectoryMode = "0750";
+          })
+          (mkIf (cfg.watchdog) {
+            ExecStartPre = "+" + pkgs.writeShellScript "configure-watchdog.sh" ''
+              ${pkgs.kmod}/bin/modprobe softdog
+              ${pkgs.coreutils}/bin/chown postgres /dev/watchdog
+            '';
+          })];
+      };
+    };
+
+    environment.systemPackages = [
+      patroni
+      pkgs.python310Packages.pysyncobj
```

Comment:
what is this needed for?

Reply:
It provides `syncobj_admin` which is useful to manage RAFT, although it makes no sense to include it if RAFT is not enabled, I'll change that.